### PR TITLE
Watcher: Properly find next valid date in cron expressions

### PR DESCRIPTION
**Pull request by spinscale**
When a list of cron expressions is provided, and one of those addresses
is already expired, the expired one will be considered as an option
instead of the valid next one.

This commit also reduces the visibility of the CronnableSchedule and
refactors a comparator to look like java 8.

### CODE REVIEW

**Comment by spinscale** [comment]
@elasticmachine retest this please

File: x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/trigger/schedule/CronnableSchedule.java [hunk 1]
```diff
@@ -37,7 +32,15 @@ public long nextScheduledTimeAfter(long startTime, long time) {
         assert time >= startTime;
         long nextTime = Long.MAX_VALUE;
         for (Cron cron : crons) {
-            nextTime = Math.min(nextTime, cron.getNextValidTimeAfter(time));
+            long nextValidTimeAfter = cron.getNextValidTimeAfter(time);
```
Comment (hub-cap) [inline x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/trigger/schedule/CronnableSchedule.java]: Im curious, does the comparator not put these at either the begin or end of the sorted list here? should we just short circuit the loop once we find one of these outliers?

File: x-pack/plugin/watcher/src/test/java/org/elasticsearch/xpack/watcher/trigger/schedule/CronScheduleTests.java [hunk 2]
```diff
@@ -54,18 +58,25 @@ public void testParseMultiple() throws Exception {
         assertThat(crons, hasItemInArray("0 0/3 * * * ?"));
     }
 
+    public void testMultipleCronsNextScheduledAfter() {
```
Comment (hub-cap) [inline x-pack/plugin/watcher/src/test/java/org/elasticsearch/xpack/watcher/trigger/schedule/CronScheduleTests.java]: should we add an expired cron entry to the list here too just to make sure?

**Review by hub-cap — APPROVED** [review]
i added 2 Qs, take them or leave them. Im +1 to this.

File: x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/trigger/schedule/CronnableSchedule.java (hunk 1 above)
Comment (spinscale) [reply x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/trigger/schedule/CronnableSchedule.java]: the cron comparator is a string based comparator (which I don't think makes a ton of sense in context of a cron expression, maybe we can just remove it), and thus the sorting could be wrong in regards to finding the next valid time.

Comment (spinscale) [reply x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/trigger/schedule/CronnableSchedule.java]: example: `0 6 9 1 1 ? 2020` and `1 5 9 1 1 ? 2020`. The sorting puts the first expression in front, but the second one is actually triggered first

File: x-pack/plugin/watcher/src/test/java/org/elasticsearch/xpack/watcher/trigger/schedule/CronScheduleTests.java (hunk 2 above)
Comment (spinscale) [reply x-pack/plugin/watcher/src/test/java/org/elasticsearch/xpack/watcher/trigger/schedule/CronScheduleTests.java]: done

**Pull request merged**